### PR TITLE
Use configured gson instance for toString

### LINKED ISSUE #768: MessageJsonHandler::toString doesn't work with custom type adapters
Hi,

I am trying to use types like `java.nio.file.Path` directly in my RPC messages. For this, we are using a custom type adapter, and registering it when building the launcher:
```java
var launcher = new Launcher.Builder<MyClient>()
      .setLocalService(myServer)
      .setRemoteInterface(MyClient.class)
      .setInput(in)
      .setOutput(out)
      .configureGson(gsonBuilder -> gsonBuilder
        .registerTypeAdapterFactory(new PathTypeAdapter.Factory())
      )
      //.traceMessages(new PrintWriter(System.out, true))  // This is one way to trigger the issue
      .create();
```

the problem is with some code in lsp4j that will try to serialize/deserialize messages with a separate instance of gson, that has not been configured with the custom type adapters.
It leads to errors like:

```text
com.google.gson.JsonIOException: Failed making field 'sun.nio.fs.UnixPath#fs' accessible; either increase its visibility or write a custom TypeAdapter for its declaring type.
	at com.google.gson.internal.reflect.ReflectionHelper.makeAccessible(ReflectionHelper.java:38)
	at com.google.gson.internal.bind.ReflectiveTypeAdapterFactory.getBoundFields(ReflectiveTypeAdapterFactory.java:287)
	at com.google.gson.internal.bind.ReflectiveTypeAdapterFactory.create(ReflectiveTypeAdapterFactory.java:130)
	at com.google.gson.Gson.getAdapter(Gson.java:546)
	at org.eclipse.lsp4j.jsonrpc.json.adapters.CollectionTypeAdapter.write(CollectionTypeAdapter.java:138)
	at org.eclipse.lsp4j.jsonrpc.json.adapters.CollectionTypeAdapter.write(CollectionTypeAdapter.java:40)
	at com.google.gson.internal.bind.TypeAdapterRuntimeTypeWrapper.write(TypeAdapterRuntimeTypeWrapper.java:70)
	at com.google.gson.internal.bind.ReflectiveTypeAdapterFactory$1.write(ReflectiveTypeAdapterFactory.java:196)
	at com.google.gson.internal.bind.ReflectiveTypeAdapterFactory$Adapter.write(ReflectiveTypeAdapterFactory.java:366)
	at com.google.gson.Gson.toJson(Gson.java:825)
	at com.google.gson.Gson.toJson(Gson.java:795)
	at com.google.gson.Gson.toJson(Gson.java:742)
	at com.google.gson.Gson.toJson(Gson.java:719)
	at org.eclipse.lsp4j.jsonrpc.json.MessageJsonHandler.toString(MessageJsonHandler.java:161)
	at org.eclipse.lsp4j.jsonrpc.TracingMessageConsumer.consumeMessageSending(TracingMessageConsumer.java:125)
	at org.eclipse.lsp4j.jsonrpc.TracingMessageConsumer.consume(TracingMessageConsumer.java:101)
```

Any use of `MessageJsonHandler.toString` is affected.

I wonder if there could be a way to either reuse the same configured gson instance everywhere, or maybe allow to also configure the one used for `toString`?

### PR DESCRIPTION
**Pull request by henryju**
This is important when users have registered custom type adapters

Fixes #768

### CODE REVIEW

**Review by pisv — COMMENTED** [review]
Overall, the patch looks reasonable to me.

It introduces a number of breaking API changes, which need to be documented in the CHANGELOG.

There are also a few stylistic nitpicks.

However, my main concern is the removal of `toString()` methods from `Message` and related classes.

This can not only create inconveniences while debugging, but also create problems with logging.

Even in `TracingMessageConsumer` itself, there are a number of places where messages are logged, which need to be changed in the same way as https://github.com/henryju/lsp4j/blob/bd47976a6a45957e5be9973e74196e1ce5e3f212/org.eclipse.lsp4j.jsonrpc/src/main/java/org/eclipse/lsp4j/jsonrpc/TracingMessageConsumer.java#L176

Basically, each of the `log` calls in LSP4J needs to be analyzed, and amended if necessary. What's more, we'll need to ensure every time, both now and in the future, that there is an accessible instance of `MessageJsonHandler` in every such case.

I don't see a good solution here, but it seems that we need to restore those 'basic' `toString()` methods, make sure they don't throw exceptions like they do currently, and try to use the new `MessageJsonHandler.toString(Object)` method for message logging where it is possible (as, for example, in `TracingMessageConsumer` itself).

WDYT?

**Comment by pisv** [comment]
To illustrate, here is how the console log currently looks after running the `IntegrationTest`:

```
Nov 08, 2023 3:45:17 PM org.eclipse.lsp4j.jsonrpc.TracingMessageConsumer consumeMessageSending
WARNING: Unmatched response message: org.eclipse.lsp4j.jsonrpc.messages.ResponseMessage@5e1f1cb
Nov 08, 2023 3:45:17 PM org.eclipse.lsp4j.jsonrpc.TracingMessageConsumer consumeMessageSending
WARNING: Unmatched response message: org.eclipse.lsp4j.jsonrpc.messages.ResponseMessage@a0b2606
Nov 08, 2023 3:45:17 PM org.eclipse.lsp4j.jsonrpc.TracingMessageConsumer consumeMessageReceiving
WARNING: Unmatched response message: {
  "jsonrpc": "2.0",
  "id": "1",
  "result": {
    "value": "BAR"
  }
}
Nov 08, 2023 3:45:17 PM org.eclipse.lsp4j.jsonrpc.TracingMessageConsumer consumeMessageReceiving
WARNING: Unmatched response message: {
  "jsonrpc": "2.0",
  "id": "1",
  "result": {
    "value": "FOO",
    "customAdapter": "/Users/vladimir/git/lsp4j/org.eclipse.lsp4j.jsonrpc"
  }
}
```

File: org.eclipse.lsp4j.jsonrpc/src/main/java/org/eclipse/lsp4j/jsonrpc/Launcher.java [hunk 1]
```diff
@@ -235,7 +235,7 @@ public static class Builder<T> {
 		protected boolean validateMessages;
 		protected Consumer<GsonBuilder> configureGson;
 		protected ClassLoader classLoader;
-		protected MessageTracer messageTracer;
+		protected PrintWriter messageTracer;
```
Comment (szarnekow) [inline org.eclipse.lsp4j.jsonrpc/src/main/java/org/eclipse/lsp4j/jsonrpc/Launcher.java:238]: This is a public class and thereby a breaking change. Is there a way to retain the MessageTracer type for this field?

File: org.eclipse.lsp4j.jsonrpc/src/main/java/org/eclipse/lsp4j/jsonrpc/json/MessageJsonHandler.java [hunk 2]
```diff
@@ -144,21 +146,27 @@ public String serialize(Message message) {
...
+	 * Perform JSON serialization of the given object using the configuration of JSON-RPC messages
 	 * enhanced with the pretty printing option.
 	 */
-	public static String toString(Object object) {
-		if (toStringInstance == null) {
-			toStringInstance = new MessageJsonHandler(Collections.emptyMap(), gsonBuilder -> {
-				gsonBuilder.setPrettyPrinting();
-			});
+	public String toString(Object object) {
+		StringWriter writer = new StringWriter();
```
[the remark this answers is not in the document]
Comment (szarnekow) [reply org.eclipse.lsp4j.jsonrpc/src/main/java/org/eclipse/lsp4j/jsonrpc/json/MessageJsonHandler.java:155]: Oh, I misread - it used to be static... Please forget what I just wrote.

File: org.eclipse.lsp4j.jsonrpc/src/main/java/org/eclipse/lsp4j/jsonrpc/Launcher.java (hunk 1 above)
Comment (pisv) [reply org.eclipse.lsp4j.jsonrpc/src/main/java/org/eclipse/lsp4j/jsonrpc/Launcher.java:238]: Unfortunately, this is not the only breaking API change here.

The culprit seems to be the `wrapMessageConsumer` method, which now requires `MessageJsonHandler` as the second parameter.

We could, of course, deprecate the old one-parameter method, but since it would no longer be called, there doesn't seem to be much value in it.

We could also deprecate the `messageTracer` field, but similarly, since it would no longer be used, would there be much value in it?

Putting the breaking change aside, the `PrintWriter` looks like a better type for the `messageTracer` to me. It follows the pattern established by other similar fields in the `Launcher.Builder`, where the type of the field matches the parameter type of the corresponding setter.

Having said that, it would be best to find a clean non-breaking solution to this problem, of course. Any suggestions would be very welcome.

Comment (pisv) [reply org.eclipse.lsp4j.jsonrpc/src/main/java/org/eclipse/lsp4j/jsonrpc/Launcher.java:238]: OK. How about this:

```java
	public static class Builder<T> {
		...
		protected PrintWriter tracer;
		protected MessageTracer messageTracer;
		...		
		public Builder<T> traceMessages(PrintWriter tracer) {
			if (tracer != null) {
				this.tracer = tracer;
				messageTracer = new MessageTracer(tracer); // fallback to old behavior
			}
			return this;
		}
		...
		public Launcher<T> create() {
			...
			MessageJsonHandler jsonHandler = createJsonHandler();
			messageTracer = new MessageTracer(tracer, jsonHandler);
			...
		}
		...
		// same as before -- no need to change anything in this method
		protected MessageConsumer wrapMessageConsumer(MessageConsumer consumer) {
			MessageConsumer result = consumer;
			if (messageTracer != null) {
				result = messageTracer.apply(consumer);
			}
			...
		}
		...
	}
```

Note that if a subclass completely overrides (i.e. replaces) the `Builder.create()` method, like `WebSocketLauncherBuilder` does in LSP4J, it will need to be modified to set the 'enhanced' `messageTracer`. Otherwise, it will fallback to the old behaviour.

File: org.eclipse.lsp4j.jsonrpc/src/main/java/org/eclipse/lsp4j/jsonrpc/TracingMessageConsumer.java [hunk 3]
```diff
@@ -55,8 +56,9 @@ public TracingMessageConsumer(
 			Map<String, RequestMetadata> sentRequests,
 			Map<String, RequestMetadata> receivedRequests,
 			PrintWriter printWriter,
+		  MessageJsonHandler jsonHandler,
```
[the remark this answers is not in the document]
Comment (henryju) [reply org.eclipse.lsp4j.jsonrpc/src/main/java/org/eclipse/lsp4j/jsonrpc/TracingMessageConsumer.java]: Fixed

File: org.eclipse.lsp4j.jsonrpc/src/test/java/org/eclipse/lsp4j/jsonrpc/test/PathTypeAdapter.java [hunk 4]
```diff
@@ -0,0 +1,41 @@
+/******************************************************************************
+ * Copyright (c) 2016 TypeFox and others.
```
[the remark this answers is not in the document]
Comment (henryju) [reply org.eclipse.lsp4j.jsonrpc/src/test/java/org/eclipse/lsp4j/jsonrpc/test/PathTypeAdapter.java]: Fixed

File: org.eclipse.lsp4j.jsonrpc/src/test/java/org/eclipse/lsp4j/jsonrpc/test/PathTypeAdapter.java [hunk 5]
```diff
@@ -0,0 +1,41 @@
...
+import com.google.gson.stream.JsonReader;
+import com.google.gson.stream.JsonToken;
+import com.google.gson.stream.JsonWriter;
+import java.io.IOException;
+import java.nio.file.Path;
+
+public class PathTypeAdapter extends TypeAdapter<Path> {
+
+  @Override
+  public void write(JsonWriter out, Path value) throws IOException {
```
[the remark this answers is not in the document]
Comment (henryju) [reply org.eclipse.lsp4j.jsonrpc/src/test/java/org/eclipse/lsp4j/jsonrpc/test/PathTypeAdapter.java]: Fixed

File: org.eclipse.lsp4j.jsonrpc/src/test/java/org/eclipse/lsp4j/jsonrpc/test/PathTypeAdapter.java [hunk 6]
```diff
@@ -0,0 +1,41 @@
...
+  @Override
+  public Path read(JsonReader in) throws IOException {
+    var peek = in.peek();
+    if (peek == JsonToken.NULL) {
+      in.nextNull();
+      return null;
+    }
+    return Path.of(in.nextString());
+  }
+}
```
[the remark this answers is not in the document]
Comment (henryju) [reply org.eclipse.lsp4j.jsonrpc/src/test/java/org/eclipse/lsp4j/jsonrpc/test/PathTypeAdapter.java]: Fixed

File: org.eclipse.lsp4j.jsonrpc/src/main/java/org/eclipse/lsp4j/jsonrpc/TracingMessageConsumer.java [hunk 7]
```diff
@@ -161,30 +165,30 @@ private String consumeMessageReceiving(Message message, Instant now, String date
...
+			String paramsJson = jsonHandler.toString(params);
 			String format = "[Trace - %s] Received request '%s - (%s)'\nParams: %s\n\n\n";
 			return String.format(format, date, method, id, paramsJson);
 		} else if (message instanceof ResponseMessage) {
 			ResponseMessage responseMessage = (ResponseMessage) message;
 			String id = responseMessage.getId();
 			RequestMetadata requestMetadata = sentRequests.remove(id);
 			if (requestMetadata == null) {
-				LOG.log(WARNING, String.format("Unmatched response message: %s", message));
+				LOG.log(WARNING, String.format("Unmatched response message: %s", jsonHandler.toString(message)));
```
[the remark this answers is not in the document]
Comment (henryju) [reply org.eclipse.lsp4j.jsonrpc/src/main/java/org/eclipse/lsp4j/jsonrpc/TracingMessageConsumer.java:176]: I fixed the one I could find, but I agree there is a risk of someone still relying on the default `toString` of a message. I will try to think about a better solution.

File: org.eclipse.lsp4j.jsonrpc/src/main/java/org/eclipse/lsp4j/jsonrpc/json/MessageJsonHandler.java [hunk 8]
```diff
@@ -11,6 +11,8 @@
  ******************************************************************************/
 package org.eclipse.lsp4j.jsonrpc.json;
 
+import com.google.gson.JsonNull;
```
[the remark this answers is not in the document]
Comment (henryju) [reply org.eclipse.lsp4j.jsonrpc/src/main/java/org/eclipse/lsp4j/jsonrpc/json/MessageJsonHandler.java]: I am not sure what are you coding guidelines: when I tried to organize imports, I found places having different orders, but I tried to fix at least the classes I touched in this PR (this is in a separate commit if you want to review).

File: org.eclipse.lsp4j.jsonrpc/src/main/java/org/eclipse/lsp4j/jsonrpc/Launcher.java (hunk 1 above)
Comment (henryju) [reply org.eclipse.lsp4j.jsonrpc/src/main/java/org/eclipse/lsp4j/jsonrpc/Launcher.java]: I made an attempt to not break any public API. The downside is that the jsonHandler will be set on the MessageTracer later on. Not sure if this could have any bad consequence.

**Comment by henryju** [comment]
I fixed most of the feedback, except the point about having no more `toString` on the various Message classes. I will have another look a bit later, to see if that would be possible to inject the jsonHandler into each of those objects when they are created.

**Review by pisv — COMMENTED** [review]
Thank you very much for your work @henryju. I have left additional comments.

File: org.eclipse.lsp4j.jsonrpc/src/main/java/org/eclipse/lsp4j/jsonrpc/json/MessageJsonHandler.java (hunk 8 above)
[the remark this answers is not in the document]
Comment (pisv) [reply org.eclipse.lsp4j.jsonrpc/src/main/java/org/eclipse/lsp4j/jsonrpc/json/MessageJsonHandler.java]: Imports can actually be a bit of a mess occasionally. Thank you for your help in cleaning it up.

File: org.eclipse.lsp4j.jsonrpc/src/main/java/org/eclipse/lsp4j/jsonrpc/TracingMessageConsumer.java [hunk 9]
```diff
@@ -72,12 +74,14 @@ public TracingMessageConsumer(
 			Map<String, RequestMetadata> sentRequests,
 			Map<String, RequestMetadata> receivedRequests,
 			PrintWriter printWriter,
+			MessageJsonHandler jsonHandler,
```
[the remark this answers is not in the document]
Comment (henryju) [reply org.eclipse.lsp4j.jsonrpc/src/main/java/org/eclipse/lsp4j/jsonrpc/TracingMessageConsumer.java:77]: By 

> class needs to fallback to the old behaviour.

do you mean it should create a new Gson instance and try to serialize it without custom adapters? I find this dangerous as it may still throw an exception as before.

[the remark this answers is not in the document]
Comment (pisv) [reply org.eclipse.lsp4j.jsonrpc/src/main/java/org/eclipse/lsp4j/jsonrpc/TracingMessageConsumer.java:77]: I think that in general we need to keep the old behavior as a fallback in cases where a `MessageJsonHandler` is unavailable for whatever reason, but make sure that possible exceptions are properly handled and no longer thrown from those `toString` methods. The goal is to avoid any breaking changes in API or behavior.

[the remark this answers is not in the document]
Comment (henryju) [reply org.eclipse.lsp4j.jsonrpc/src/main/java/org/eclipse/lsp4j/jsonrpc/TracingMessageConsumer.java:127]: I made another attempt. I didn't like the fact that the jsonHandler could be null on the `MessageTracer`, so I went for a deprecation approach.

[the remark this answers is not in the document]
Comment (pisv) [reply org.eclipse.lsp4j.jsonrpc/src/main/java/org/eclipse/lsp4j/jsonrpc/TracingMessageConsumer.java:127]: If you ask me, I liked the previous approach better. To me, the problem looks like a `progressive enhancement` of the old behaviour in cases where a `MessageJsonHandler` is available. As I have mentioned, there might still be cases where it will not be available, so I can see no reason why we should require it in a constructor. This is my 'informed opinion' so to speak. Nevertheless, I really appreciate your effort.

File: org.eclipse.lsp4j.jsonrpc/src/main/java/org/eclipse/lsp4j/jsonrpc/MessageTracer.java [hunk 10]
```diff
@@ -11,31 +11,34 @@
...
  * https://microsoft.github.io/language-server-protocol/inspector/
  */
 public class MessageTracer implements Function<MessageConsumer, MessageConsumer> {
 	private final PrintWriter printWriter;
+	private final MessageJsonHandler jsonHandler;
 	private final Map<String, RequestMetadata> sentRequests = new HashMap<>();
 	private final Map<String, RequestMetadata> receivedRequests = new HashMap<>();
 
-	MessageTracer(PrintWriter printWriter) {
+	MessageTracer(PrintWriter printWriter, MessageJsonHandler jsonHandler) {
```
Comment (pisv) [inline org.eclipse.lsp4j.jsonrpc/src/main/java/org/eclipse/lsp4j/jsonrpc/MessageTracer.java:34]: This is also a breaking API change.

Comment (henryju) [reply org.eclipse.lsp4j.jsonrpc/src/main/java/org/eclipse/lsp4j/jsonrpc/MessageTracer.java:34]: The constructor is package private, so I would say this is not a breaking API change

File: org.eclipse.lsp4j.jsonrpc/src/main/java/org/eclipse/lsp4j/jsonrpc/TracingMessageConsumer.java (hunk 9 above)
[the remark this answers is not in the document]
Comment (henryju) [reply org.eclipse.lsp4j.jsonrpc/src/main/java/org/eclipse/lsp4j/jsonrpc/TracingMessageConsumer.java:127]: Since `MessageTracer` constructor is package private, I take for granted that it is not instantiated/extended outside lsp4j. So changing it is fine.
Regarding `TracingMessageConsumer`, I deprecated the old constructors, but keep them working as before using a fresh `MessageJsonHandler` instance.
If you think the two use cases are valid on the long run, I can remove deprecation tags.
The last "issue" is the removal of `toString` on the various messages. I don't like reverting to the old behavior, since this could be really creating bugs on user side. What about providing `toString` with the best information we could, but not relying on gson serialization?

File: org.eclipse.lsp4j.jsonrpc/src/main/java/org/eclipse/lsp4j/jsonrpc/MessageTracer.java (hunk 10 above)
Comment (pisv) [reply org.eclipse.lsp4j.jsonrpc/src/main/java/org/eclipse/lsp4j/jsonrpc/MessageTracer.java:34]: My bad. You are right, of course. But I'd still prefer a setter to the constructor in this case.

**Comment by pisv** [comment]
To sum up and hopefully clarify, here are what I'd like to see in this PR.

Thanks to the prompt by @szarnekow  and the following changes by @henryju, I have been persuaded that there is a clean non-breaking solution to this problem.

So, one of the main goals should be to preserve all of the existing API elements and augment the existing behaviour rather than trying to replace it completely.

To put it in a perspective, there were no issues reported about this problem since 2019 when the `TracingMessageConsumer` was introduced. This seems to indicate that for most of the LSP4J clients the current implementation works just fine.

Therefore, let's keep the current API and implementation to avoid breaking any of the existing clients, and just augment it with the enhanced implementation where it is possible, using the old one as a fallback.

So, let's first reinstate all of the removed API elements and their behavior, including the static `MessageJsonHandler.toString` method, `Method.toString`, `CancelParams.toString`, `ResponseError.toString`.

Let's rename the instance `MessageJsonHandler.toString` method to e.g. `format`.

Let's use a setter for passing a `MessageJsonHandler` to both `MessageTracer` and `TracingMessageConsumer` without touching any of the existing constructors as it is the least invasive approach and works just fine in this case.

When no `MessageJsonHandler` is set on a `TracingMessageConsumer`, let's fallback to the existing behaviour.

That would be a clean non-breaking solution I'd like to see.

As a bonus, let's make sure that possible exceptions are properly handled in the  existing static `MessageJsonHandler.toString` method, so that it becomes more resilient if some of the required type adapters are missing.

However, I'd consider this bonus part as optional for this PR.

**Comment by henryju** [comment]
I won't say I am super convinced by keeping dangerous `toString` implementations, but I will respect your decision. I have some other things to do today, so I will come back to this next week.

**Comment by pisv** [comment]
> I won't say I am super convinced by keeping dangerous `toString` implementations

One one hand, at least it will not make things any worse for existing clients.

With your current proposal, clients would start to see something like

```
WARNING: Unmatched response message: Message{id=1}
```

instead of

```
WARNING: Unmatched response message: {
  "jsonrpc": "2.0",
  "id": "1",
  "result": {
    "value": "FOO"
  }
}
```

I don't think it would be fine for most purposes.

On the other hand, we can try to make the 'dangerous' `toString` implementations less dangerous:

> As a bonus, let's make sure that possible exceptions are properly handled in the existing static MessageJsonHandler.toString method, so that it becomes more resilient if some of the required type adapters are missing.

File: org.eclipse.lsp4j.jsonrpc/src/main/java/org/eclipse/lsp4j/jsonrpc/Launcher.java [hunk 11]
```diff
@@ -294,7 +299,7 @@ public Builder<T> validateMessages(boolean validate) {
 		
 		public Builder<T> traceMessages(PrintWriter tracer) {
 			if (tracer != null) {
-				this.messageTracer = new MessageTracer(tracer);
```
Comment (pisv) [inline org.eclipse.lsp4j.jsonrpc/src/main/java/org/eclipse/lsp4j/jsonrpc/Launcher.java:297]: With your current approach, we'd still need to keep this line with the one-arg constructor as a fallback. Otherwise, an existing subclass that completely overrides `Builder.create()` would have no `messageTracer` set, which is a breaking change. Similarly, we'd need to make the new two-arg constructor public so that such subclasses could reset the default `messageTracer` in their `create()` implementations. That's why I much prefer your previous approach with `MessageTracer.setJsonHandler`. It is so much less of a hassle.

**Comment by pisv** [comment]
For the sake of convenience, I have pushed what I'd consider a clean patch against `main` that reflects the points I have depicted above. (It doesn't include the bonus part.)

If you don't mind, let's use it as a baseline for further discussion and development.

**Comment by pisv** [comment]
Let's see if I can come up with something along the lines sketched out in https://github.com/eclipse-lsp4j/lsp4j/issues/768#issuecomment-1807164354 to make the 'dangerous' `toString` methods less dangerous...

File: org.eclipse.lsp4j.jsonrpc/src/main/java/org/eclipse/lsp4j/jsonrpc/messages/ToStringBuilder.java
[the remark this answers is not in the document]
Comment (jonahgraham) [reply org.eclipse.lsp4j.jsonrpc/src/main/java/org/eclipse/lsp4j/jsonrpc/messages/ToStringBuilder.java:1]: As lsp4j.jsonrpc is at the root of all dependency chains I think having it just in this new place is fine. And it is certainly fine to defer to a future PR

[the remark this answers is not in the document]
Comment (cdietrich) [reply org.eclipse.lsp4j.jsonrpc/src/main/java/org/eclipse/lsp4j/jsonrpc/messages/ToStringBuilder.java:1]: to my knowledge its not
https://github.com/eclipse-lsp4j/lsp4j/blob/main/org.eclipse.lsp4j.jsonrpc/build.gradle

**Comment by pisv** [comment]
@jonahgraham Could you please review this PR with a fresh pair of eyes?

We tried to enhance the existing implementation without any breaking changes by passing a specific `MessageJsonHandler` to the `TracingMessageConsumer` and `Message` so that the static `MessageJsonHandler.toString(Object)` method is called only as a fallback now.

If all else fails, `Message.toString()`, `CancelParams.toString()`, and `ResponseError.toString()` will fallback to an alternative implementation based on a `ToStringBuilder`.

As far as I can see, the main potential issue with the current patch is that existing reflective code might be broken by the addition of the `Message.jsonHandler` field. (There is a more detailed [comment](https://github.com/eclipse-lsp4j/lsp4j/pull/772#discussion_r1395709514) above about this issue.) Apart from that, it should cause no issues for existing clients.

**Comment by pisv** [comment]
Regarding the potential issue with existing reflective code, I now think that it would probably be sufficient to just mention it in the CHANGELOG as a potential breaking change, but I would still appreciate an extra pair of eyes to review this PR.

It would be great if we could merge it before the 0.22.0 release.

**Comment by jonahgraham** [comment]
I added it to the 0.22.0 milestone. It fell off the bottom of my todo list, I will schedule to look at it sometime before the 0.22.0 release. See conversation about timing in #732

**Comment by pisv** [comment]
> I added it to the 0.22.0 milestone. It fell off the bottom of my todo list, I will schedule to look at it sometime before the 0.22.0 release. See conversation about timing in #732

OK. Thank you!

File: org.eclipse.lsp4j.jsonrpc/src/main/java/org/eclipse/lsp4j/jsonrpc/messages/Message.java [hunk 12]
```diff
@@ -11,16 +11,34 @@
...
 
+import com.google.gson.JsonIOException;
+
 /**
  * A general message as defined by JSON-RPC. The language server protocol always
  * uses "2.0" as the jsonrpc version.
  */
 public abstract class Message {
 
+	private transient MessageJsonHandler jsonHandler;
```
[the remark this answers is not in the document]
Comment (jonahgraham) [reply org.eclipse.lsp4j.jsonrpc/src/main/java/org/eclipse/lsp4j/jsonrpc/messages/Message.java:28]: I am not really concerned by this - but I have added an entry to the Changelog that mentions what to do in this case.

**Review by jonahgraham — APPROVED** [review]
This looks fine to me. I am about to push a rebased squashed version with the above mentioned changelog entry too to ensure that it builds cleanly.

**Comment by jonahgraham** [comment]
>  next year

next *week :-)

File: CHANGELOG.md [hunk 13]
```diff
@@ -6,6 +6,9 @@ Fixed issues: <https://github.com/eclipse-lsp4j/lsp4j/milestone/29?closed=1>
 
 Breaking API changes:
 
+  * The Message class now has a new transient field, `jsonHandler` to enable the `toString` implementation to properly format messages when custom type adapters are in used.
```
Comment (pisv) [inline CHANGELOG.md:9]:
```suggestion
  * The Message class now has a new transient field, `jsonHandler`, to enable the `toString` implementation to properly format messages when custom type adapters are used.
```

File: CHANGELOG.md [hunk 14]
```diff
@@ -6,6 +6,9 @@ Fixed issues: <https://github.com/eclipse-lsp4j/lsp4j/milestone/29?closed=1>
...
+    * For consumers that have their own custom serializers they may need to be updated to ensure that transient fields are skipped, for example by using [`Modifier.isTransient`](https://docs.oracle.com/en/java/javase/11/docs/api/java.base/java/lang/reflect/Modifier.html#isTransient(int))
```
Comment (pisv) [inline CHANGELOG.md:10]:
```suggestion
    * For consumers that have their own custom serializers or other reflective message processors they may need to be updated to ensure that transient fields are skipped, for example by using [`Modifier.isTransient`](https://docs.oracle.com/en/java/javase/11/docs/api/java.base/java/lang/reflect/Modifier.html#isTransient(int))
```

**Review by pisv — COMMENTED** [review]
@jonahgraham Thank you very much for your review and approval.

May I suggest a couple of minor enhancements to the changelog entry?

**Comment by jonahgraham** [comment]
Thanks everyone. This is now committed and the next release of LSP4J (expected next week) will contain this improvement.
